### PR TITLE
Implements group_by "time_zone" and "time_zone_utc" in nb_inventory

### PR DESCRIPTION
**Pull request by hyberdk**
## Related Issue
Implements: "[Feature]: nb_inventory add group_by time_zone and time_zone_utc #788"

## New Behavior

In the inventory source module, you can now group_by time_zone and time_zone_utc

We needed a way to group our devices by time zone or by UTC time to easily schedule playbooks during the night.
...

## Contrast to Current Behavior

Currently we had to run the playbook on all hosts and then filter out the devices in the wrong time zone. This was proned to errors and accidental tasks being run that were not supposed to.
...

## Discussion: Benefits and Drawbacks

Im sure that other people that uses AWX/Tower to push nightly tasks to devices will like this change as it simplifies the way you can filter devices in a particular time zone.

I dont see many drawbacks as if you dont use the time_zone or time_zone_utc group by it does not affect the existing setup. It will also detect if its not being used and not build the dicts hence no process overhead.

Only drawback I see is that it needs the pytz module to run.

grouping looks like this:

```
    "time_zone_Africa_Johannesburg": {
        "hosts": [
            "zahpf-core01",
            "zahpf-dmvpn01",
            "zancc-core01",
            "zancc-dmvpn01",
            "zazbx-dmvpn01",
            "zazcg-core01",
            "zazcg-dmvpn01"
        ]
    },
    "time_zone_Africa_Nairobi": {
        "hosts": [
            "keloy-dmvpn01",
            "kenbo-dmvpn01"
        ]
    },

-----

    "time_zone_utc_minus_7": {
        "hosts": [
            "uspdxapw-core01",
            "usqie-dmvpn01",
            "usrgj-as01",
            "usrgj-dr01",
            "usrtp-dr01",
            "uswc7-sw01"
        ]
    },
    "time_zone_utc_plus_1": {
        "hosts": [
            "iecas-dmvpn01",
            "iecos-core01",
            "ptldb-dr01",
            "ptldb-mdf01-srv-sw01",
            "ptldb-n-ff-i01-as01",
            "ptplg-dmvpn01"
        ]
    },

```

...

## Changes to the Documentation

not much, just added the options in group_by. I think its pretty much self explanatory 

...

## Proposed Release Note Entry

nb_inventory - implements group by time_zone and time_zone_utc #788
...

## Double Check

* [x] I have read the comments and followed the [CONTRIBUTING.md](https://github.com/netbox-community/ansible_modules/blob/devel/CONTRIBUTING.md).
* [x] I have explained my PR according to the information in the comments or in a linked issue.
* [x] My PR targets the `devel` branch.

### CODE REVIEW

**Comment by sc68cal** [comment]
The one thing I see is that `time_zone` and `time_zone_utc` fields are a bit ambiguous, I understand what you are trying to do, but I think maybe a little time needs to be spent on the `time_zone_utc` field name, since that's a little confusing to me.

**Comment by sc68cal** [comment]
Off the top of my head, here's my suggestion:

`time_zone_name`

`time_zone_utc_offset` 

That would make things more clear.

**Comment by hyberdk** [comment]
> time_zone_utc_offset

Perhaps just utc_offset would make sense then?

I guess time_zone is a bit redundant if we add the offset.

Also I did not find a good place to document the feature

**Comment by sc68cal** [comment]
`time_zone` and `utc_offset` works for me. Much clearer what those values are.

As for documentation, most of it is autogenerated from the code via sphinx. So as long as you add information as comments in the actual code, sphinx should pick it up and add it to the docs.

**Comment by sc68cal** [comment]
Basically, add an example of sorting your inventory by time_zone and utc_offset to the end of the `EXAMPLE` variable in https://github.com/ansible-collections/dellemc.os10/blob/master/roles/os10_acl/tasks/main.yml  and sphinx should pick it up and add it to https://netbox-ansible-collection.readthedocs.io/en/latest/plugins/nb_inventory_inventory.html#examples

**Comment by hyberdk** [comment]
So I  renamed time_zone_utc to utc_offset and wrote a little in the example to make it clear.. Hopefully it will also pass linting now..

Please let me know if there are other things you guys think should be changed (or added).

**Comment by hyberdk** [comment]
> This is the error: `ERROR: Found 1 import issue(s) on python 3.8 which need to be resolved: [128](https://github.com/netbox-community/ansible_modules/pull/798/checks#step:8:129) ERROR: plugins/inventory/nb_inventory.py:318:0: traceback: ModuleNotFoundError: No module named 'pytz'`
> 
> It seems a defined dependency has stopped using `pytz` as a dependency, so we should update the `pyproject.toml` file with an explicit dependency on `pytz`. I'll see if I can get that done tonight.

ahh okay.. its strange I dont see that in the output (https://github.com/netbox-community/ansible_modules/runs/6895917693?check_suite_focus=true)

but awesome if you can have a look at it tonight! Thanks a lot!

File: plugins/inventory/nb_inventory.py [hunk 1]
```diff
@@ -245,6 +245,16 @@
 
 # has_primary_ip is a useful way to filter out patch panels and other passive devices
 
+# If you group by time_zone, utc_offset etc. it will group devices in ansible groups depending on time zone configured on site.
```
[the remark this answers is not in the document]
Comment (sc68cal) [reply plugins/inventory/nb_inventory.py:248]:
```yaml
plugin: netbox.netbox.nb_inventory
api_endpoint: http://localhost:8000
group_by:
  - time_zone
```

```yaml
plugin: netbox.netbox.nb_inventory
api_endpoint: http://localhost:8000
group_by:
  - utc_offset
```

**Comment by hyberdk** [comment]
> I'd also like to see some unit tests added and sample data added to validate that this works and we never break it by accident. Great job so far, just a little bit more to go

Im not the sharpest for pytest, so please bear with me.. when i clone a new copy and run `pytest -vv` this is what I get:

```
(venv) eslau@N503476:~/blah/ansible_collections/netbox/netbox$ pytest -vv
========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.8.10, pytest-7.1.2, pluggy-1.0.0 -- /home/eslau/blah/ansible_collections/netbox/netbox/venv/bin/python3
cachedir: .pytest_cache
rootdir: /home/eslau/blah/ansible_collections/netbox/netbox
collected 0 items / 2 errors

================================================================================================= ERRORS =================================================================================================
_______________________________________________________________________ ERROR collecting tests/unit/inventory/test_nb_inventory.py _______________________________________________________________________
<frozen importlib._bootstrap>:908: in _find_spec
    ???
E   AttributeError: '_AnsibleCollectionFinder' object has no attribute 'find_spec'

During handling of the above exception, another exception occurred:
tests/unit/inventory/test_nb_inventory.py:19: in <module>
    from ansible_collections.netbox.netbox.tests.test_data import load_test_data
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:971: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:910: in _find_spec
    ???
<frozen importlib._bootstrap>:884: in _find_spec_legacy
    ???
venv/lib/python3.8/site-packages/ansible/utils/collection_loader/_collection_finder.py:222: in find_module
    return _AnsibleCollectionLoader(fullname=fullname, path_list=path)
venv/lib/python3.8/site-packages/ansible/utils/collection_loader/_collection_finder.py:314: in __init__
    self._subpackage_search_paths = self._get_subpackage_search_paths(self._candidate_paths)
venv/lib/python3.8/site-packages/ansible/utils/collection_loader/_collection_finder.py:607: in _get_subpackage_search_paths
    collection_meta = _get_collection_metadata(collection_name)
venv/lib/python3.8/site-packages/ansible/utils/collection_loader/_collection_finder.py:1064: in _get_collection_metadata
    raise ValueError('collection metadata was not loaded for collection {0}'.format(collection_name))
E   ValueError: collection metadata was not loaded for collection netbox.netbox
___________________________________________________________________ ERROR collecting tests/unit/module_utils/test_netbox_base_class.py ___________________________________________________________________
<frozen importlib._bootstrap>:908: in _find_spec
    ???
E   AttributeError: '_AnsibleCollectionFinder' object has no attribute 'find_spec'

During handling of the above exception, another exception occurred:
tests/unit/module_utils/test_netbox_base_class.py:17: in <module>
    from ansible_collections.netbox.netbox.plugins.module_utils.netbox_utils import (
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:971: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:910: in _find_spec
    ???
<frozen importlib._bootstrap>:884: in _find_spec_legacy
    ???
venv/lib/python3.8/site-packages/ansible/utils/collection_loader/_collection_finder.py:222: in find_module
    return _AnsibleCollectionLoader(fullname=fullname, path_list=path)
venv/lib/python3.8/site-packages/ansible/utils/collection_loader/_collection_finder.py:314: in __init__
    self._subpackage_search_paths = self._get_subpackage_search_paths(self._candidate_paths)
venv/lib/python3.8/site-packages/ansible/utils/collection_loader/_collection_finder.py:607: in _get_subpackage_search_paths
    collection_meta = _get_collection_metadata(collection_name)
venv/lib/python3.8/site-packages/ansible/utils/collection_loader/_collection_finder.py:1064: in _get_collection_metadata
    raise ValueError('collection metadata was not loaded for collection {0}'.format(collection_name))
E   ValueError: collection metadata was not loaded for collection netbox.netbox
======================================================================================== short test summary info =========================================================================================
ERROR tests/unit/inventory/test_nb_inventory.py - ValueError: collection metadata was not loaded for collection netbox.netbox
ERROR tests/unit/module_utils/test_netbox_base_class.py - ValueError: collection metadata was not loaded for collection netbox.netbox
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=========================================================================================== 2 errors in 0.50s ============================================================================================
(venv) eslau@N503476:~/blah/ansible_collections/netbox/netbox$
```
Im not sure why that is exactly. 

I was also looking into the integration tests, but the documentation is referring to a file "integration-tests.yml"  https://github.com/netbox-community/ansible_modules/blob/de4637081c14f0acb3a5cac63f380a8474b39b47/CONTRIBUTING.md?plain=1#L296 

but I cannot find it and I dont think Im supposed to run the "runme.sh" script manually..

Hope it makes sense.

**Comment by sc68cal** [comment]
I would take a look at our CI configuration to see how to get your environment set up. I think there's a bunch of things that `poetry` handles to get things set up, so try running the commands in sequence, that the CI system runs.

https://github.com/netbox-community/ansible_modules/blob/devel/.github/workflows/main.yml

**Comment by hyberdk** [comment]
> I'll see if I can get that done tonight.

Hi @rodvand, I was wondering if you had a chance to look at this.. Im scratching my head, I tried to add pytz to [requirements.txt](https://github.com/netbox-community/ansible_modules/blob/devel/requirements.txt) and also to the [pyproject.toml](https://github.com/netbox-community/ansible_modules/blob/devel/pyproject.toml) file, but I cannot make it succeed the `ansible-test units`  

if you know where to update the dependency advise would be much appreciated.

**Comment by rodvand** [comment]
> > I'll see if I can get that done tonight.
> 
> Hi @rodvand, I was wondering if you had a chance to look at this.. Im scratching my head, I tried to add pytz to [requirements.txt](https://github.com/netbox-community/ansible_modules/blob/devel/requirements.txt) and also to the [pyproject.toml](https://github.com/netbox-community/ansible_modules/blob/devel/pyproject.toml) file, but I cannot make it succeed the `ansible-test units`
> 
> if you know where to update the dependency advise would be much appreciated.

You might have to add something like https://github.com/netbox-community/ansible_modules/blob/8ef1932f2a40d1d913429e939d8da0abcd927c9d/plugins/module_utils/netbox_utils.py#L24 to try/except the import.

**Comment by sc68cal** [comment]
This looks good, I just want to see an example of configuring the plugin with these new directives, as per my review comments.

**Comment by hyberdk** [comment]
Hi @sc68cal,

Ohh yeah, sorry about that, I have added https://github.com/netbox-community/ansible_modules/pull/798/commits/469d1bb2abbdaa48050be995f448171d4b334681, I hope that was what you were looking for, else let me know.

Esben

**Review by sc68cal — APPROVED** [review]
Looks good to me

**Pull request merged**